tests/test_physics/test_bargmann.py

```diff
@@ -51,3 +57,103 @@ def test_bargmann_numpy_transformation():
     """Tests that the numpy option of the bargmann method of State works correctly"""
     transformation = Ggate(1)
     assert all(isinstance(thing, np.ndarray) for thing in transformation.bargmann(numpy=True))
+
+
+def test_abc_contraction_2mode_psi_U():
+    "tests that the abc contraction works for U|psi>"
+    psi = Gaussian(2)
+    U = Ggate(2)
+
+    A1, b1, c1 = psi.bargmann()  # out1ket, out2ket
+    A2, b2, c2 = U.bargmann()  # out1ket, out2ket, in1ket, in2ket
+
+    A_abc, b_abc, c_abc = contract_two_Abc((A1, b1, c1), (A2, b2, c2), (0, 1), (2, 3))
+    A_mm, b_mm, c_mm = (psi >> U).bargmann()
+
+    assert np.allclose(A_abc, A_mm)
+    assert np.allclose(b_abc, b_mm)
+    assert np.allclose(abs(c_abc), abs(c_mm))
+
+
+def test_abc_contraction_2mode_rho_phi():
+    "tests that the abc contraction works for rho >> phi"
+    rho = Gaussian(2) >> Attenuator([0.1, 0.2]) >> Ggate(2) >> Attenuator([0.4, 0.9])
+    phi = Ggate(2) >> Attenuator([0.3, 0.4]) >> Ggate(2)
+    # out1bra, out2bra, out1ket, out2ket
+    A1, b1, c1 = rho.bargmann()
+    # out1bra, out2bra, in1bra, in2bra, out1ket, out2ket, in1ket, in2ket
+    A2, b2, c2 = phi.bargmann()
+
+    A_abc, b_abc, c_abc = contract_two_Abc((A1, b1, c1), (A2, b2, c2), (0, 1, 2, 3), (2, 3, 6, 7))
+
+    A_mm, b_mm, c_mm = (rho >> phi).bargmann()
+
+    assert np.allclose(A_abc, A_mm)
+    assert np.allclose(b_abc, b_mm)
+    assert np.allclose(c_abc, c_mm)
+
+
+def test_abc_contraction_3mode_rho_2mode_U():
+    "tests that the abc contraction works for U rho U_dagger"
+    rho = Gaussian(3) >> Attenuator([0.1, 0.2, 0.4]) >> Ggate(3) >> Attenuator([0.4, 0.5, 0.9])
+    U = Ggate(2)
+
+    # out1bra, out2bra, out3bra, out1ket, out2ket, out3ket
+    A1, b1, c1 = rho.bargmann()
+    # out1ket, out2ket, in1ket, in2ket
+    A2, b2, c2 = U.bargmann()
+
+    A_abc, b_abc, c_abc = contract_two_Abc(
+        (A2, b2, c2), (A1, b1, c1), (2, 3), (4, 5)
+    )  # left in out1ket_U, out2ket_U, out1bra_rho, out2bra_rho, out3bra_rho, out1ket_rho
+
+    A_abc, b_abc, c_abc = contract_two_Abc(
+        (A_abc, b_abc, c_abc),
+        (math.conj(A2), math.conj(b2), math.conj(c2)),
+        (3, 4),
+        (2, 3),
+    )  # left in out1ket_U, out2ket_U, out1bra_rho, out1ket_rho, out1bra_U, out2bra_U
+
+    A_abc, b_abc, c_abc = reorder_abc((A_abc, b_abc, c_abc), (2, 4, 5, 3, 0, 1))
+
+    A_mm, b_mm, c_mm = (rho >> U[1, 2]).bargmann()
+
+    assert np.allclose(A_abc, A_mm)
+    assert np.allclose(b_abc, b_mm)
+    assert np.allclose(c_abc, c_mm)
+
+
+def test_Bargmann_2mode_psi_U():
```

Comment:
I think that tests like this one (which are effectively testing `Bargmann` and its matmul) should be moved to a file `test_representations.py`, which does not exist at the moment.

Specifically, I'm talking about this test and the next 2

Reply:
Is this another bug? If so, we should add more tests to `test_representations.py` to make sure it's all fixed

```
from mrmustard.physics.representations import Bargmann
import numpy as np

a_vacuum = np.zeros((2, 2))
b_vacuum = np.zeros((2,))
c_vacuum = 1
vacuum = Bargmann(a_vacuum, b_vacuum, c_vacuum)

gamma = 1 + 1j
a_dgate = np.zeros((2, 2))
b_dgate = np.array([gamma, -np.conj(gamma)])
c_dgate = np.exp(-abs(gamma)**2/2)
dgate = Bargmann(a_dgate, b_dgate, c_dgate)

ret = vacuum[0] @ dgate[1]
ret.b
# expected [1-1j, 1+1j], found [0+0j, 1+1j]
```

Reply:
I don't think you're doing what you think you're doing

Reply:
Quite possible!